cmd/snap-update-ns/utils.go

```diff
@@ -45,6 +47,84 @@ var (
 	sysFchown  = syscall.Fchown
 )
 
+// Create directories for all but the last segments and return the file
+// descriptor to the leaf directory. This function is a base for secure
+// variants of mkdir, touch and symlink.
+func secureMkPrefix(segments []string, perm os.FileMode, uid, gid int) (int, error) {
+	logger.Debugf("secure-mk-prefix %q %v %d %d -> ...", segments, perm, uid, gid)
+
+	// Declare var and don't assign-declare below to ensure we don't swallow
+	// any errors by mistake.
+	var err error
+	var fd int
+
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
+
+	// Open the root directory and start there.
+	fd, err = sysOpen("/", openFlags, 0)
+	if err != nil {
+		return -1, fmt.Errorf("cannot open root directory: %v", err)
+	}
+	if len(segments) > 1 {
+		defer sysClose(fd)
+	}
+
+	// Process all but the last segment.
+	for i := range segments[:len(segments)-1] {
+		fd, err = secureMkDir(fd, segments, i, perm, uid, gid)
+		if err != nil {
+			return -1, err
+		}
+		// Keep the final FD open (caller needs to close it).
+		if i < len(segments)-2 {
+			defer sysClose(fd)
+		}
+	}
+
+	logger.Debugf("secure-mk-prefix %q %v %d %d -> %d", segments, perm, uid, gid, fd)
+	return fd, nil
+}
+
+// secureMkdir creates a directory at i-th entry of absolute path represented
+// by segments. This function can be used to construct subsequent elements of
+// the constructed path.
```

Comment:
Maybe add a note that `secureMkDir` returns a file descriptor to newly created segment and the it's -1 on errors.